### PR TITLE
use shared_ptr for unit_ptr

### PR DESCRIPTION
**Pull request by gfgtdf**
This make the code a bit simpler, furthermore it
allows us to use weak_ptr to untsi shodul we ever
need that.

### CODE REVIEW

**Comment by gfgtdf** [comment]
okay this doesn't work compltey yet

**Comment by CelticMinstrel** [comment]
I don't think it's correct to replace unit references with unit pointers. You should use pointer or reference based on whether the function is equipped to deal with a null. The ones you changed from reference to pointer should probably be changed back.

I didn't look at this in detail yet, but did you use `std::enable_shared_from_this` for unit? If so I would recommend exposing the `shared_from_this()` function publicly (perhaps by overloading the address operator to return it).

**Comment by gfgtdf** [comment]
hmm converting to unit_ptr to refernces and back is not omething i realyl liek to be honest. I'm thinking about adding a new class 'shared_reference' that's basicially  shared_ptr wrapper that cannot be assigned nullptr.

**Comment by CelticMinstrel** [comment]
> hmm converting to unit_ptr to refernces and back is not omething i realyl liek to be honest.

Why not? If it works for normal pointers, why shouldn't we allow it to work for shared pointers? I think a shared_reference class would be overkill for this, honestly. That said... I can't say it's _wrong_. There _is_ one minor downside to taking the reference, which is that it's not an owning reference. I think that usually won't be an issue, but... with a shared_reference type class you could make it an owning reference. At the very least, I would prefer a shared_reference over replacing all those references by pointers.

**Comment by gfgtdf** [comment]
okay there were some cases in the naimation code that converted between unit_ptr and unit* so i had to use shared_from_this anyways if i didnt want to to a really large refactor.. But it should fork fine now

**Comment by CelticMinstrel** [comment]
I definitely like overloading operator& to return shared_from_this(), but I suppose I'm probably in the minority there…? Either way, get_shared_ptr() is perfectly reasonable too. Maybe a tad long, but whatever.

**Comment by gfgtdf** [comment]
> I definitely like overloading operator& to return shared_from_this(),

there are major problems with overloading operator &, for example in case where unit is incomplete, it will still return a noprmal pointer so the result of `&u` depends on whether unit.hpp was included or not.

**Comment by stevecotton** [comment]
Doesn't build with GCC 10, it needs an `#include <stdexcept>` in src/utils/shared_reference.hpp.

Also, there's no documentation for the shared_reference class, other than the commit message.

**Comment by CelticMinstrel** [comment]
The shared_reference class looks fairly self-explanatory to me - it's literally a non-nullable shared_ptr.

Again I might be in the minority, but I probably would have written `operator&` instead of `operator shared_ptr`. I realize that there's a common wisdom that overloading the address-of operator is "bad", but my understanding is it's bad because you can change the semantics of it. Overloading it to return a `shared_ptr` does not change the semantics, so in my opinion it's a reasonable thing to do.

> there are major problems with overloading operator &, for example in case where unit is incomplete, it will still return a noprmal pointer so the result of `&u` depends on whether unit.hpp was included or not.

This _is_ a legitimate issue that I hadn't thought of. However, I don't think it would allow incorrect code to compile in many cases (if any). If you're taking the address of a unit, chances are you're going to pass it to a function that takes a `unit_ptr` (or return it as a `unit_ptr`), so that will fail to compile because `unit*` is not implicitly convertible to `unit_ptr`. If you're not doing that, you're probably doing something that otherwise requires `unit` to be complete, such as accessing a member of it.

So I would classify it as a minor problem, rather than a major problem, though there could be some case I haven't thought of that could change my mind.

**Comment by soliton-** [comment]
I think it would go a long way to simply add a single line description instead of requiring everyone to read the implementation.

File: src/units/unit.hpp [hunk 1]
```diff
@@ -210,7 +210,7 @@ class unit : public std::enable_shared_from_this<unit>
 
 	unit_ptr clone() const
 	{
-		return unit_ptr(new unit(*this));
+		return unit_ptr(std::shared_ptr<unit>(new unit(*this)));
```
Comment (AI0867) [inline src/units/unit.hpp:213]: `return unit_ptr(std::make_shared<unit>(*this))`?

Comment (gfgtdf) [reply src/units/unit.hpp:213]: yes i should do this, the reason i added that line is that i temporarily, replaced unit_ptr with a wrapper for std::shared_ptr<unit> without the constructor that takes 1 unit* to detect all cases hen unit* is converted into unit_ptr

**Pull request merged**